### PR TITLE
20240327-tls-int-overflows

### PR DESCRIPTION
**Pull request by douzzer**
`src/internal.c`: mitigations for potential integer overflows in figuring allocation sizes.

tested with `wolfssl-multi-test.sh ... super-quick-check`

see ZD#17733

### CODE REVIEW

File: src/internal.c [hunk 1]
```diff
@@ -10561,8 +10561,7 @@ static WC_INLINE int GrowOutputBuffer(WOLFSSL* ssl, int size)
 #else
     const byte align = WOLFSSL_GENERAL_ALIGNMENT;
 #endif
-    int newSz = size + ssl->buffers.outputBuffer.idx +
-                ssl->buffers.outputBuffer.length;
+    size_t newSz;
```
Comment (SparkiDev) [inline src/internal.c:10564]: Don't use size_t unless we have to.
Would word32 do here?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/internal.c (hunk 1 above)
Comment (douzzer) [reply src/internal.c:10564]: not as far as I can figure it, because the whole point is to catch int wrap on the arg to `XMALLOC()`, which is `size_t`.  I think we really need `size_t` specifically, whatever its width.

anyway the old practice around `size_t` went out the window with the pedantic `-Wconversion` cleanup about a year ago.  see ed11669f3c, 4b59588cf3, etc.

File: src/internal.c [hunk 2]
```diff
@@ -33377,7 +33388,7 @@ static int DoSessionTicket(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
     #if defined(HAVE_ECC) || defined(HAVE_CURVE25519) || defined(HAVE_CURVE448)
         word32 exportSz;
     #endif
-        int    sendSz;
+        size_t sendSz;
```
Comment (dgarske) [inline src/internal.c:33391]: Why was this changed?

File: src/internal.c [hunk 3]
```diff
@@ -10578,7 +10577,15 @@ static WC_INLINE int GrowOutputBuffer(WOLFSSL* ssl, int size)
         align *= 2;
 #endif
 
-    tmp = (byte*)XMALLOC(newSz + align, ssl->heap, DYNAMIC_TYPE_OUT_BUFFER);
+    if (! WC_SAFE_SUM_WORD32(ssl->buffers.outputBuffer.idx,
+                             ssl->buffers.outputBuffer.length, newSz))
+        return BUFFER_E;
+    if (! WC_SAFE_SUM_WORD32(newSz, (word32)size, newSz))
```
Comment (dgarske) [inline src/internal.c:10583]: `c:\jenkins\workspace\prb-windows-test-v2\src\internal.c(10583): warning C4701: potentially uninitialized local variable 'newSz' used [C:\jenkins\workspace\PRB-windows-test-v2\wolfssl.vcxproj]
Lib`

**Review by dgarske — CHANGES_REQUESTED** [review]

**Review by dgarske — APPROVED** [review]
